pkg/builder/builder_steps.go

```diff
@@ -22,38 +22,99 @@ import (
 	"io/ioutil"
 	"os"
 	"path"
+	"reflect"
 	"strings"
 
 	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/scylladb/go-set/strset"
 
 	"github.com/apache/camel-k/pkg/apis/camel/v1alpha1"
+	"github.com/apache/camel-k/pkg/util/maven"
 	"github.com/apache/camel-k/pkg/util/tar"
 
 	yaml2 "gopkg.in/yaml.v2"
 
 	"github.com/pkg/errors"
-
-	"github.com/apache/camel-k/pkg/util/maven"
 )
 
-// GenerateProject --
-func GenerateProject(ctx *Context) error {
+var StepsByID = make(map[string]Step)
+
+func init() {
+	RegisterSteps(Steps)
+}
+
+type steps struct {
+	GenerateProject      Step
+	InjectDependencies   Step
+	SanitizeDependencies Step
+	ComputeDependencies  Step
+	StandardPackager     Step
+	IncrementalPackager  Step
+}
+
+var Steps = steps{
+	GenerateProject: NewStep(
+		"project/generate",
+		ProjectGenerationPhase,
+		generateProject,
+	),
+	InjectDependencies: NewStep(
+		"project/inject-dependencies",
+		ProjectGenerationPhase+1,
+		injectDependencies,
+	),
+	SanitizeDependencies: NewStep(
+		"project/sanitize-dependencies",
+		ProjectGenerationPhase+2,
+		sanitizeDependencies,
+	),
+	ComputeDependencies: NewStep(
+		"build/compute-dependencies",
+		ProjectBuildPhase,
+		computeDependencies,
+	),
+	StandardPackager: NewStep(
+		"packager",
+		ApplicationPackagePhase,
+		standardPackager,
+	),
+	IncrementalPackager: NewStep(
+		"packager/incremental",
+		ApplicationPackagePhase,
+		incrementalPackager,
+	),
+}
+
+func RegisterSteps(steps interface{}) {
+	v := reflect.ValueOf(steps)
+	for i := 0; i < v.NumField(); i++ {
+		if step, ok := v.Field(i).Interface().(Step); ok {
+			RegisterStep(step)
+		}
+	}
+}
+
+func RegisterStep(steps ...Step) {
+	for _, step := range steps {
+		StepsByID[step.ID()] = step
```

Comment:
now that the id has a real meaning wondering if we need to add some tests/check to avoid overriding a step by mistake

Reply:
Yes I thought about it but got somehow a bit lazy 🙈. I guess we can fail-fast and panic because step overriding should not happen.

Reply:
yep but a test would help too :)

Reply:
I've added panicking and unit testing, and updated the builder integration tests to cover all the packages.